editor/plugins/embedded_process.cpp

```diff
@@ -0,0 +1,353 @@
+void EmbeddedProcess::_timer_embedding_timeout() {
+	_try_embed_process();
+}
+
+void EmbeddedProcess::_draw() {
+	if (_focused_process_id == _current_process_id && has_focus() && _focus_style_box.is_valid()) {
+		Size2 size = get_size();
+		Rect2 r = Rect2(Point2(), size);
+		_focus_style_box->draw(get_canvas_item(), r);
+	}
+}
+
+void EmbeddedProcess::_check_mouse_over() {
+	// This method checks if the mouse is over the embedded process while the current application is focused.
+	// The goal is to give focus to the embedded process as soon as the mouse hovers over it,
+	// allowing the user to interact with it immediately without needing to click first.
+	if (!is_visible_in_tree() || !_embedding_completed || !_application_has_focus || !_window || !_window->has_focus() || Input::get_singleton()->is_mouse_button_pressed(MouseButton::LEFT) || Input::get_singleton()->is_mouse_button_pressed(MouseButton::RIGHT)) {
+		return;
+	}
+
+	bool focused = has_focus();
+
+	// Not stealing focus from a textfield.
+	if (!focused && get_viewport()->gui_get_focus_owner() && get_viewport()->gui_get_focus_owner()->is_text_field()) {
+		return;
+	}
+
+	Vector2 mouse_position = DisplayServer::get_singleton()->mouse_get_position();
+	Rect2i window_rect = get_screen_embedded_window_rect();
+	if (!window_rect.has_point(mouse_position)) {
+		return;
+	}
+
+	// Don't grab the focus if mouse over another window.
+	DisplayServer::WindowID window_id_over = DisplayServer::get_singleton()->get_window_at_screen_position(mouse_position);
+	if (window_id_over > 0 && window_id_over != _window->get_window_id()) {
+		return;
+	}
+
+	// When we already have the focus and the user moves the mouse over the embedded process,
+	// we just need to refocus the process.
+	if (focused) {
+		_queue_update_embedded_process();
+	} else {
+		grab_focus();
+		queue_redraw();
+	}
+}
+
+void EmbeddedProcess::_check_focused_process_id() {
+	OS::ProcessID process_id = DisplayServer::get_singleton()->get_focused_process_id();
+	if (process_id != _focused_process_id) {
+		_focused_process_id = process_id;
+		if (_focused_process_id == _current_process_id) {
+			// The embedded process got the focus.
+			if (has_focus()) {
+				// Redraw to updated the focus style.
+				queue_redraw();
+			} else {
+				grab_focus();
+			}
+		} else if (has_focus()) {
+			release_focus();
+		}
+	}
+}
+
+void EmbeddedProcess::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("embed_process", "process_id"), &EmbeddedProcess::embed_process);
+	ClassDB::bind_method(D_METHOD("reset"), &EmbeddedProcess::reset);
+	ClassDB::bind_method(D_METHOD("set_embedding_timeout", "timeout"), &EmbeddedProcess::set_embedding_timeout);
+	ClassDB::bind_method(D_METHOD("get_embedding_timeout"), &EmbeddedProcess::get_embedding_timeout);
+	ClassDB::bind_method(D_METHOD("is_embedding_completed"), &EmbeddedProcess::is_embedding_completed);
+	ClassDB::bind_method(D_METHOD("is_embedding_in_progress"), &EmbeddedProcess::is_embedding_in_progress);
```

Comment:
What are these bound for?

Reply:
At first, I tough maybe this control could be public but maybe later. I removed it.